File: CS/WinForms.Client/ClientSecurity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Threading.Tasks;

namespace WinForms.Client {
    static class ClientSecurity {
        static readonly HttpClient HttpClient = new HttpClient();
        //
        static string BaseUrl {
            get { return System.Configuration.ConfigurationManager.AppSettings["baseURL"]; }
        }
        static string AuthUrl {
            get { return System.Configuration.ConfigurationManager.AppSettings["authURL"]; }
        }
        static string PermissionUrl {
            get { return System.Configuration.ConfigurationManager.AppSettings["checkPermissionURL"]; }
        }
        public static string ODataUrl {
            get { return System.Configuration.ConfigurationManager.AppSettings["baseODataUrl"]; }
        }
        //
        public static void Initialize() {
            HttpClient.BaseAddress = new Uri(BaseUrl);
        }
        static string AuthToken;
        public static void SetAuthToken(string authToken) {
            AuthToken = authToken;
        }
        public static async Task<string> Authenticate(string userName, string password) {
            HttpClient.DefaultRequestHeaders.Accept.Clear();
            HttpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            var responseMessage = await HttpClient.PostAsJsonAsync(AuthUrl, new {
                UserName = userName,
                Password = password
            }).ConfigureAwait(false);
            if(responseMessage.IsSuccessStatusCode) {
                string token = await responseMessage.Content.ReadAsStringAsync()
                    .ConfigureAwait(false);
                return token.Replace("\"", "").Trim();
            }
            return null;
        }
        public static async Task<bool> EnsureEditPermissions(string objectTypeFullName, string objectKey) {
            HttpClient.DefaultRequestHeaders.Accept.Clear();
            HttpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            HttpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", AuthToken);
            var responseMessage = await HttpClient.GetAsync(PermissionUrl + $"?typeName={objectTypeFullName}&objectKey={objectKey}")
                .ConfigureAwait(false);
            if(responseMessage.IsSuccessStatusCode) {
                string result = await responseMessage.Content.ReadAsStringAsync()
                    .ConfigureAwait(false);
                return bool.Parse(result);
            }
            return false;
        }
        public static void EnsureAuthorization(Microsoft.OData.Client.BuildingRequestEventArgs args) {
            args.Headers.Add("Authorization", "Bearer " + AuthToken); // Add authentification token
        }
    }
}

File: VB/WinForms.Client/ClientSecurity.vb
Imports System
Imports System.Linq
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports System.Net.Http.Json

Namespace WinForms.Client

    Friend Module ClientSecurity

        Private ReadOnly HttpClient As HttpClient = New HttpClient()

        '
        Private ReadOnly Property BaseUrl As String
            Get
                Return Configuration.ConfigurationManager.AppSettings("baseURL")
            End Get
        End Property

        Private ReadOnly Property AuthUrl As String
            Get
                Return Configuration.ConfigurationManager.AppSettings("authURL")
            End Get
        End Property

        Private ReadOnly Property PermissionUrl As String
            Get
                Return Configuration.ConfigurationManager.AppSettings("checkPermissionURL")
            End Get
        End Property

        Public ReadOnly Property ODataUrl As String
            Get
                Return Configuration.ConfigurationManager.AppSettings("baseODataUrl")
            End Get
        End Property

        '
        Public Sub Initialize()
            HttpClient.BaseAddress = New Uri(BaseUrl)
        End Sub

        Private AuthToken As String

        Public Sub SetAuthToken(ByVal authToken As String)
            ClientSecurity.AuthToken = authToken
        End Sub

        Public Async Function Authenticate(ByVal userName As String, ByVal password As String) As Task(Of String)
            Call HttpClient.DefaultRequestHeaders.Accept.Clear()
            Call HttpClient.DefaultRequestHeaders.Accept.Add(New MediaTypeWithQualityHeaderValue("application/json"))
            Dim responseMessage = Await HttpClient.PostAsJsonAsync(AuthUrl, New With {userName, password}).ConfigureAwait(False)
            If responseMessage.IsSuccessStatusCode Then
                Dim token As String = Await responseMessage.Content.ReadAsStringAsync().ConfigureAwait(False)
                Return token.Replace("""", "").Trim()
            End If

            Return Nothing
        End Function

        Public Async Function EnsureEditPermissions(ByVal objectTypeFullName As String, ByVal objectKey As String) As Task(Of Boolean)
            Call HttpClient.DefaultRequestHeaders.Accept.Clear()
            Call HttpClient.DefaultRequestHeaders.Accept.Add(New MediaTypeWithQualityHeaderValue("application/json"))
            HttpClient.DefaultRequestHeaders.Authorization = New AuthenticationHeaderValue("Bearer", AuthToken)
            Dim responseMessage = Await HttpClient.GetAsync(PermissionUrl & $"?typeName={objectTypeFullName}&objectKey={objectKey}").ConfigureAwait(False)
            If responseMessage.IsSuccessStatusCode Then
                Dim result As String = Await responseMessage.Content.ReadAsStringAsync().ConfigureAwait(False)
                Return Boolean.Parse(result)
            End If

            Return False
        End Function

        Public Sub EnsureAuthorization(ByVal args As Microsoft.OData.Client.BuildingRequestEventArgs)
            args.Headers.Add("Authorization", "Bearer " & AuthToken) ' Add authentification token
        End Sub
    End Module
End Namespace
